Handle situation where logfile can't be created.

aymanhab opened the pull request:
Fixes issue #0

### Brief summary of changes
Handle the case of logfile that can't be opened.

### Testing I've completed
None
### Looking for feedback on...
What should be the behavior on failure? Before this PR, GUI fails to launch, we can have GUI add its own sink in a different folder that GUI decides on (e.g. Resources folder which we know is writable, other folder common to Netbeans, /AppData, ... or under home directory though I don't like sprinkling files all over the file system)
### CHANGELOG.md (choose one)

- no need to update because...
- updated...

The Doxygen for this PR can be viewed at http://myosin.sourceforge.net/?C=N;O=D; click the folder whose name is this PR's number.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/opensim-org/opensim-core/2797)
<!-- Reviewable:end -->

aymanhab commented:
@chrisdembia this is a strawman attempt mainly to see if the GUI would launch on Mac with this rather defensive and surgical change. To address in general we could bubble an exception but we still need to answer the question of what to do while keeping the exception handling localized.

chrisdembia commented:
We could rely on this code for determining where to write the log file: https://github.com/sago007/PlatformFolders/blob/master/sago/platform_folders.cpp

chrisdembia commented on OpenSim/Common/Logger.cpp:
```diff
@@ -150,6 +150,16 @@ void Logger::addFileSink(const std::string& filepath) {
+        ifs.close();
+    } else {
+        // show message:
+        std::cout << "Can't open file " << filepath << "Log will not be created." << std::endl;
```
Should use the logger: `log_warn(...)`.

We could add a function like `Logger::fileSinkExists()` or `Logger::hasFileSink()` that the GUI could invoke to determine if the GUI should log to some user data directory.

I think it's also fine for there to be no logging file when using the GUI—the content is in the Messages window, anyway.

aymanhab replied:
@chrisdembia The function addFileSink() is invoked from inside the constructor of Logger so it's not clear if any other logging is possible at this point, including log_cout or log_warn 😕 

I think we can make the GUI add its own FileSink as needed but no logging from GUI for now (if the file can't be created) seems less catastrophic than the current behavior.

chrisdembia replied:
That's a good point but `addFileSink()` could be invoked by a user after the constructor has completed, in which case we would want `log_warn()` to be used.

aymanhab commented:
@chrisdembia since we're using the singleton pattern to create/access the Logger, I moved the function outside the constructor and updated to use the logging syntax.

aymanhab commented:
> > Reviewed 2 of 2 files at r2.
> > _[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/2797)_ status: all files reviewed, 1 unresolved discussion (waiting on @aymanhab)
> 
> _[OpenSim/Common/Logger.cpp, line 159 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/2797#-M9ZT6NZ8BCB6CPIF4FF:-M9ZT6NZ8BCB6CPIF4FG:b-yh61di) ([raw file](https://github.com/opensim-org/opensim-core/blob/aad14db3924cf7af55b52d6745eeb0aad073f79a/OpenSim/Common/Logger.cpp#L159)):_
> 
> > ```c++
> >     } else {
> >         // show message:
> >          warn("Can't open file {} for writing. Log will not be created.", 
> > ```
> 
> Consider being a little more specific: `File sink will not be created.`

@chrisdembia I think from a user perspective "Sink" makes no sense while log does. Thoughts?

aymanhab commented:
> We could rely on this code for determining where to write the log file: https://github.com/sago007/PlatformFolders/blob/master/sago/platform_folders.cpp

Thanks fr finding this @chrisdembia 
There's similar logic in the application 
https://github.com/opensim-org/opensim-gui/blob/5aaa56d214331f9730384133631cea4e45cf6737/Gui/opensim/utils/src/org/opensim/utils/TheApp.java#L189
which seems like better suited for Platform specific code than low level API. 

It's also possible that users run API programs from different folders to keep separate log files and if we force a common location at the API we'd work against them, while an application can set its own convention.

chrisdembia commented:
> @chrisdembia I think from a user perspective "Sink" makes no sense while log does. Thoughts?

Perhaps "Log file" then. "Log" on its own could mean that messages will not be written to the console.

aymanhab commented:
Done, thanks @chrisdembia

chrisdembia commented on OpenSim/Common/Logger.cpp:
```diff
@@ -150,6 +149,17 @@ void Logger::addFileSink(const std::string& filepath) {
+        ifs.close();
+    } else {
+        // show message:
+         warn("Can't open file {} for writing. Log file will not be created.", 
```
Can you make the style of this warning consistent with the one above?

aymanhab replied:
Not sure what specifically you want changed, I took a stab at it. As a one liner, feel free to change as well as you see fit. Thank you.

chrisdembia commented on OpenSim/Common/Logger.cpp:
```diff
@@ -150,6 +149,18 @@ void Logger::addFileSink(const std::string& filepath) {
              "removeFileSink() first.", m_filesink->filename());
```
@aymanhab thanks for adding the quotes below. You mentioned that "file sink" is not clear to users; perhaps the usage of "file sink" in this warning should be changed to "log file" for consistency.

aymanhab commented:
Thanks @chrisdembia

aymanhab commented:
@chrisdembia this rather innocent looking change is causing the test case testCommandLineInterface to fail because of failing to open a log file. Any clues about what the test case assumes? Debugging the test case is problematic because it uses "system" methods that the debugger can't step through 😨

chrisdembia commented:
I'll take a look.

chrisdembia commented:
@aymanhab I don't have that issue locally; do you encounter this issue locally?

chrisdembia commented:
@aymanhab I think the issue is multiple processes (tests running in parallel) trying to open the same file. Instead of trying to open the file ourselves, I suggest we catch an exception thrown by `std::make_shared<spdlog::sinks::basic_file_sink_mt>(filepath);`

aymanhab commented:
@chrisdembia I don't have the issue locally, and I don't understand why this behavior has changed to begin with but I can add a catch as you suggested

aymanhab commented:
This seems to have taken care of the issue somehow, thanks for the suggestion @chrisdembia

chrisdembia commented:
I think `ifs.good()` was false because the file was already open, rather than because of a lack of write privileges.
